example/example_tests.py

```diff
@@ -0,0 +1,211 @@
+
+        z = longitudinal_velocity * self.t
+        ux = self._ux(z - 0.5 * self.time_step)
+        uz = (self.mean_lorentz_factor ** 2 - 1 - ux ** 2) ** 0.5
+        x = (
+            ux[0] / p.lorentz_factor * self.time_step / 2
+            + np.cumsum(ux / p.lorentz_factor) * self.time_step
+        )
+        y = self._y(z)
+        uy = self._uy(z)
+
+        p.track = Track(x, y, z, ux, uy, uz, p.weight)
+        self.particles.append(p)
+
+    def create_particles(self):
+        deviation = self.energy_spread * np.random.randn(self.number_of_particles)
+        lorentz_factors = self.mean_lorentz_factor + deviation
+
+        for lf in lorentz_factors:
+            p = Particle(lorentz_factor=lf)
+            self._add_particle(p)
+
+        return self
+
+    def __getitem__(self, p):
+        return self.particles[p]
+
+    # Calculating the particles orbits (simplified)
+    def _ux(self, z):
+        val = self.strength * np.sin(2 * np.pi * z)
+        val *= (z > 0) * (z < 1.5) * z / 1.5 + (z > 1.5)
+        val *= (z > self.number_of_periods - 1.5) * (z < self.number_of_periods) * (
+            self.number_of_periods - z
+        ) / 1.5 + (z < self.number_of_periods - 1.5)
+        return val
+
+    @staticmethod
+    def _y(z):
+        return np.zeros_like(z)
+
+    @staticmethod
+    def _uy(z):
+        return np.zeros_like(z)
+
+    def energy_analytical(self):
+        return (
+            self.number_of_particles
+            * self.central_wavenumber
+            * J_in_um
+            * (7 * np.pi / 24)
+            / 137.0
+            * self.strength ** 2
+            * (1 + self.strength ** 2 / 2)
+            * self.number_of_periods
+        )
+
+
+@attr.s
+class Particle:
+    lorentz_factor = attr.ib()
+    track = attr.ib(default=None)
+    weight = attr.ib(default=1.0)
+
+
+@attr.s
+class Track:
+    x = attr.ib()
+    y = attr.ib()
+    z = attr.ib()
+    ux = attr.ib()
+    uy = attr.ib()
+    uz = attr.ib()
+    w = attr.ib()
+
+    def aslist(self):
+        return [self.x, self.y, self.z, self.ux, self.uy, self.uz, self.w]
+
+
+u = Undulator()
+
+
+@attr.s
+class Grid:
+    from_k = attr.ib(default=0.02 * u.central_wavenumber)
+    to_k = attr.ib(default=1.1 * u.central_wavenumber)
+    resolution_k = 512
+
+    from_theta = attr.ib(default=0.0)
+    to_theta = attr.ib(default=2.0 / u.mean_lorentz_factor)
+    resolution_theta = 256
+
+    from_phi = attr.ib(default=0.0)
+    to_phi = attr.ib(default=2 * np.pi)
+    resolution_phi = 36
+
+    def to_list(self):
+        return [
+            (self.from_k, self.to_k),
+            (self.from_theta, self.to_theta),
+            (self.from_phi, self.to_phi),
+            (self.resolution_k, self.resolution_theta, self.resolution_phi),
+        ]
+
+
+@attr.s
+class CalcInput:
+    grid = attr.ib(default=Grid().to_list())
+    timeStep = attr.ib(default=u.time_step)
+    dtype = attr.ib(default="float")
+    native = attr.ib(default=True)
```

Comment:
I have even added a new one now, called `Resolution`. I prefer to work by composing various small objects, where each of them does only one thing. While it may seem like an overkill now, I find this leads to more flexibility and code reuse down the line. In particular, `CalcInput` is only necessary to make the connection between the `Undulator` and `SynchRad`. It will look better once `SynchRad` will be able to accept directly a parameter class.

Reply:
Basically wherever you have a dictionary with a fixed set of keys that you don't iterate over, that should be a class ;)